deluge/ui/gtkui/gtkui.py

```diff
@@ -239,35 +240,51 @@ def nsapp_open_file(osxapp, filename):
 
         # Twisted catches signals to terminate, so have it call a pre_shutdown method.
         reactor.addSystemEventTrigger("before", "shutdown", self.pre_shutdown)
+        reactor.addSystemEventTrigger("before", "gtkui_close", self.close)
+
         reactor.callWhenRunning(self._on_reactor_start)
+        self.closing = False
+
+        def gtkui_sigint_handler(num, frame):
+            log.debug("SIGINT signal caught - firing event: 'gtkui_close'")
+            reactor.callLater(0, reactor.fireSystemEvent, 'gtkui_close')
+
+        signal.signal(signal.SIGINT, gtkui_sigint_handler)
 
         # Initialize gdk threading
         gtk.gdk.threads_enter()
         reactor.run()
-        self.shutdown()
-        gtk.gdk.threads_leave()
-
-    def shutdown(self, *args, **kwargs):
-        log.debug("gtkui shutting down..")
-
-        component.stop()
 
         # Process any pending gtk events since the mainloop has been quit
         while gtk.events_pending():
-            gtk.main_iteration(0)
+            gtk.main_iteration_do(block=False)
+            time.sleep(0.0001)
```

Comment:
Not sure, just found this: http://stackoverflow.com/questions/2844902/python-four-steps-setup-with-progressbars
I was going to ask whether this code is even necessary at all...?